src/DefaultBuilder/src/WebApplicationBuilder.cs

```diff
@@ -101,11 +100,66 @@ internal WebApplicationBuilder(Assembly? callingAssembly, string[]? args = null)
+                {
+                    builder.Properties[key] = value;
+                }
+            });
+
             // We call ConfigureWebHostDefaults AGAIN because config might be added like "ForwardedHeaders_Enabled"
             // which can add even more services. If not for that, we probably call _hostBuilder.ConfigureWebHost(ConfigureWebHost)
             // instead in order to avoid duplicate service registration.
             _hostBuilder.ConfigureWebHostDefaults(ConfigureWebHost);
-            return _builtApplication = new WebApplication(_hostBuilder.Build());
+
+            // Copy the services that were added via WebApplicationBuilder.Services into the final IServiceCollection
+            _hostBuilder.ConfigureServices((context, services) =>
+            {
+                // We've only added services configured by the GenericWebHostBuilder and WebHost.ConfigureWebDefaults
+                // at this point. HostBuilder news up a new ServiceCollection in HostBuilder.Build() we haven't seen
+                // until now, so we cannot clear these services even though some are redundant because
+                // we called ConfigureWebHostDefaults on both the _deferredHostBuilder and _hostBuilder.
+
+                // Ideally, we'd only call _hostBuilder.ConfigureWebHost(ConfigureWebHost) instead of
+                // _hostBuilder.ConfigureWebHostDefaults(ConfigureWebHost) to avoid some duplicate service descriptors,
+                // but we want to add services in the WebApplicationBuilder constructor so code can inspect
+                // WebApplicationBuilder.Services. At the same time, we want to be able which services are loaded
+                // to react to config changes (e.g. ForwardedHeadersStartupFilter).
+                foreach (var s in _services)
+                {
+                    services.Add(s);
+                }
+
+                // Add any services to the user visible service collection so that they are observable
+                // just in case users capture the Services property. Orchard does this to get a "blueprint"
+                // of the service collection
+
+                // Clear the original collection.
+                _services.Clear();
+
+                // Drop the reference to the existing collection and set the inner collection
+                // to the new one. This allows code that has references to the service collection to still function.
+                _services.InnerCollection = services;
+            });
+
+            // Run the other callbacks on the final host builder
+            _deferredHostBuilder.RunDeferredCallbacks(_hostBuilder);
+
+            _builtApplication = new WebApplication(_hostBuilder.Build());
+
+            // Make builder.Configuration match the final configuration. To do that
+            // we clear the sources and add the built configuration as a source
+            ((IConfigurationBuilder)Configuration).Sources.Clear();
```

Comment:
So we just lose the added configuration if a deferred callback adds a source? I'm hoping to eliminate the deferred callback completely though. We've already inlined most of them.

Reply:
You can't, look at the tests, it has no way of getting visibility into what happens after these callbacks.

Reply:
For some reason I was thinking this was pre-build even though Build() is only a couple line up.